pkg/agent/controller/networkpolicy/l7engine/l7visbility_watcher.go

```diff
@@ -0,0 +1,397 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package l7engine
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+
+	"antrea.io/antrea/pkg/agent/controller/trafficcontrol"
+)
+
+var (
+	L7mutex sync.Mutex
+)
+
+const (
+	watcherName = "L7Watcher"
+)
+
+type l7Watcher struct {
+	tcController  *trafficcontrol.Controller
+	clientset     kubernetes.Interface
+	l7Reconciler  *Reconciler
+	labelsToTCMap map[string][]string
+}
+
+func NewL7Watcher(tcController *trafficcontrol.Controller, clientset kubernetes.Interface) *l7Watcher {
+	l7Reconciler := NewReconciler()
+	labelsToTCMap := make(map[string][]string)
+	return &l7Watcher{
+		tcController:  tcController,
+		clientset:     clientset,
+		l7Reconciler:  l7Reconciler,
+		labelsToTCMap: labelsToTCMap,
+	}
+}
+
+func (l7w *l7Watcher) Run(stopCh <-chan struct{}) {
+	klog.InfoS("Starting", "Watcher", watcherName)
+	defer klog.InfoS("Shutting down", "Watcher", watcherName)
+	// Create a watcher for pods and namespaces
+	podWatcher, err := l7w.clientset.CoreV1().Pods("").Watch(context.TODO(), metav1.ListOptions{
+		FieldSelector: fields.Everything().String(),
+	})
```

Comment:
there are a couple of major issues here:
- we never call `Watch` directly to watch for changes using the K8s API. Instead we use the informer pattern, with a local cache, which puts less load on the K8s apiserver.
- we have a rule that we do not watch for _all_ Pods resources in the Agent, to avoid putting too much pressure on the K8s apiserver. The only acceptable option is to watch Pods local to the Node, using the appropriate filter. We have an example here: https://github.com/antrea-io/antrea/blob/81ebd61938c8273eef9bc12eef3d7c3cf6c819ae/cmd/antrea-agent/agent.go#L333-L339.

The good news is that I do not believe you actually need to watch all Pods in the cluster, since based on my understanding, your feature only cares about Pods local to the Node.

Reply:
+1, it should use the localPodInformer like other modules.